types/module/module.go

```diff
@@ -299,19 +304,38 @@ func (m *Manager) RegisterServices(cfg Configurator) {
 	}
 }
 
-// InitGenesis performs init genesis functionality for modules. Exactly one
-// module must return a non-empty validator set update to correctly initialize
-// the chain.
+func (m *Manager) UnmarshalGenesis(cdc codec.JSONCodec, genesisData map[string]json.RawMessage) map[string]proto.Message {
+	var wg sync.WaitGroup
+	genesisStates := map[string]proto.Message{}
+	for _, moduleName := range m.OrderInitGenesis {
+		if genesisData[moduleName] == nil {
+			continue
+		}
+
+		wg.Add(1)
+		go func(moduleName string) {
+			defer wg.Done()
+			genesisStates[moduleName] = m.Modules[moduleName].UnmarshalGenesis(cdc, genesisData[moduleName])
```

Comment:
There is concurrent access to `genesisStates`, which might trip off any race detectors. While in practice this should be OK since each goroutine accesses it's own unique key in `genesisStates`, we might wanna make this thread-safe. Perhaps `Manager` could have a mutex.

Reply:
This wasn't safe as written before! As genesisStates is a hashmap, with grow operations =p

The mutex operations were definitely needed for safety